include/osquery/http_client.h

```diff
@@ -0,0 +1,438 @@
+  void postResponseHandler(boost_system::error_code const& ec);
+
+  void closeSocket();
+
+ private:
+  Options client_options_;
+  boost_asio::io_service ios_;
+  boost_asio::ip::tcp::resolver r_;
+  boost_asio::ip::tcp::socket sock_;
+  boost_asio::deadline_timer timer_;
+  std::shared_ptr<ssl_stream> ssl_sock_;
+  boost_system::error_code ec_;
+  bool ssl_connection = false;
+  static const long SHORT_READ_ERROR = 0x140000dbL;
+  /// Setting the default port to squid proxy default port
+  static const int PROXY_DEFAUT_PORT = 3128;
+  static const int HTTP_DEFAULT_PORT = 80;
+  static const int HTTPS_DEFAULT_PORT = 443;
+};
+
+/**
+ * @brief HTTP request class
+ *
+ * This class is inherited from implemention(boost.beast) request class.
+ * It extends the functionality via providing uri parsing.
+ *
+ */
+template <typename T>
+class HTTP_Request : public T {
+ public:
+  /**
+   * @brief HTTP request header helper class
+   *
+   * constructor of this class takes (name, value)
+   * pair, which is used to set the request header
+   * of a http request with the help of overloaded
+   * function 'operator<<' of HTTP_Request class.
+   */
+  struct Header {
+    Header(const std::string& name, const std::string& value)
+        : name_(name), value_(value) {}
+
+   private:
+    std::string name_;
+    std::string value_;
+    friend class HTTP_Request<T>;
+  };
+
+ public:
+  HTTP_Request(const std::string& url = std::string()) : uri_(url) {}
+
+  /// Returns the host part of a uri.
+  boost::optional<std::string> remoteHost() {
+    return uri_.host().size() ? uri_.host() : boost::optional<std::string>();
+  }
+
+  /// Returns the port part of a uri.
+  boost::optional<std::string> remotePort() {
+    return uri_.port().size() ? uri_.port() : boost::optional<std::string>();
+  }
+
+  /// Returns the path part of a uri.
+  boost::optional<std::string> remotePath() {
+    return uri_.path().size() ? uri_.path() : boost::optional<std::string>();
+  }
+
+  /// Returns the protocol part of a uri. E.g. 'http' or 'https'
+  boost::optional<std::string> protocol() {
+    return uri_.scheme().size() ? uri_.scheme()
+                                : boost::optional<std::string>();
+  }
+
+  /// overloaded operator to set header of an http request
+  HTTP_Request& operator<<(const Header& h) {
+    this->T::set(h.name_, h.value_);
+    return *this;
+  }
+
+  /// uri can also be set via this method, useful for redirected request.
+  void uri(const std::string& url) {
+    uri_ = url;
+  }
+
+ private:
+  boost::network::uri::uri uri_;
+};
+
+/**
+ * @brief HTTP response class
+ *
+ * This class is inherited from implementation(boost.beast) http response class.
+ * This class gives convinent access to some functionality of implemention
+ * specific http response class.
+ *
+ */
+template <typename T>
+class HTTP_Response : public T {
+ public:
+  HTTP_Response() {}
+  HTTP_Response(const T& resp) : T(resp) {}
```

Comment:
One comment on my own code  :)
need to change constructor to - 
HTTP_Response(T&& resp) : T(std::move(resp)) {} 
in-order to achieve zero-copy here  "Response(resp.release())"